Consider using TTL more extensively in k8s executor

## New feature

Currently `ttlSecondsAfterFinished` is an optional config option for k8s executor which adds metadata to jobs/pods which tells the k8s control plane to kill resources upon completion / failure after that many seconds.

However, the nextflow monitor pool loop will actively delete resources some amount of time post completion, this appears to be variable from my observation and if you have lots of fast running processes this can add up to lots of resources for the loop to monitor slowing the main process down. Also, if the control plane kills a job before nextflow tries to this will lead to an error like this:

```
Oct-06 18:32:43.421 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Unexpected error in tasks monitor pool loop
org.codehaus.groovy.runtime.InvokerInvocationException: nextflow.k8s.client.K8sResponseException: Request DELETE /apis/batch/v1/namespaces/<NAMESPACE>/jobs
/nf-62e16f1931488513c9732bfd1e7d5718-f9506 returned an error code=404

  {
      "kind": "Status",
      "apiVersion": "v1",
      "metadata": {
          
      },
      "status": "Failure",
      "message": "jobs.batch \"nf-62e16f1931488513c9732bfd1e7d5718-f9506\" not found",
      "reason": "NotFound",
      "details": {
          "name": "nf-62e16f1931488513c9732bfd1e7d5718-f9506",
          "group": "batch",
          "kind": "jobs"
      },
      "code": 404
  }

        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:348)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:328)
        at groovy.lang.MetaClassImpl.doInvokeMethod(MetaClassImpl.java:1333)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1088)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1007)
        at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:645)
        at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:628)
        at org.codehaus.groovy.runtime.InvokerHelper.invokeMethodSafe(InvokerHelper.java:82)
        at nextflow.processor.TaskPollingMonitor$_start_closure2.doCall(TaskPollingMonitor.groovy:323)
        at nextflow.processor.TaskPollingMonitor$_start_closure2.call(TaskPollingMonitor.groovy)
        at groovy.lang.Closure.run(Closure.java:505)
        at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: nextflow.k8s.client.K8sResponseException: Request DELETE /apis/batch/v1/namespaces/<NAMESPACE>/jobs/nf-62e16f1931488513c9732bfd1e7d5718-f9506 re
turned an error code=404
```

In the log which puts the pipeline into a state where it will no longer spawn new jobs / pods but also not finish with an error.

I suggest that you consider refactoring some of the k8s functionality to make better use of k8s features (such as TTL), better utilise the control plane to cheaply achieve many of the same features, or at the very least better handle the case where a resource no longer exists when nextflow tries to delete it.

--- a/plugins/nf-k8s/src/main/nextflow/k8s/K8sTaskHandler.groovy
+++ b/plugins/nf-k8s/src/main/nextflow/k8s/K8sTaskHandler.groovy
@@ -416,8 +416,10 @@ class K8sTaskHandler extends TaskHandler implements FusionAwareTask {
 
     @Override
     boolean checkIfCompleted() {
-        if( !podName ) throw new IllegalStateException("Missing K8s ${resourceType.lower()} name - cannot check if complete")
-        def state = getState()
+        if( !podName )
+            throw new IllegalStateException("Missing K8s ${resourceType.lower()} name - cannot check if complete")
+
+        final state = getState()
         if( state && state.terminated ) {
             if( state.nodeTermination instanceof NodeTerminationException ||
                 state.nodeTermination instanceof PodUnschedulableException ) {
@@ -441,8 +443,8 @@ class K8sTaskHandler extends TaskHandler implements FusionAwareTask {
                 task.stderr = errorFile
             }
             status = TaskStatus.COMPLETED
-            savePodLogOnError(task)
-            deletePodIfSuccessful(task)
+            saveJobLogOnError(task)
+            deleteJobIfSuccessful(task)
             updateTimestamps(state.terminated as Map)
             determineNode()
             return true
@@ -451,7 +453,7 @@ class K8sTaskHandler extends TaskHandler implements FusionAwareTask {
         return false
     }
 
-    protected void savePodLogOnError(TaskRun task) {
+    protected void saveJobLogOnError(TaskRun task) {
         if( task.isSuccess() )
             return
 
@@ -491,26 +493,21 @@ class K8sTaskHandler extends TaskHandler implements FusionAwareTask {
      */
     @Override
     protected void killTask() {
+        if( !podName )
+            return
+
         if( cleanupDisabled() )
             return
         
-        if( podName ) {
-            log.trace "[K8s] deleting ${resourceType.lower()} name=$podName"
-            if ( useJobResource() )
-                client.jobDelete(podName)
-            else
-                client.podDelete(podName)
-        }
-        else {
-            log.debug "[K8s] Invalid delete action"
-        }
+        log.trace "[K8s] deleting ${resourceType.lower()} name=$podName"
+        delete0(podName)
     }
 
     protected boolean cleanupDisabled() {
         !k8sConfig.getCleanup()
     }
 
-    protected void deletePodIfSuccessful(TaskRun task) {
+    protected void deleteJobIfSuccessful(TaskRun task) {
         if( !podName )
             return
 
@@ -521,10 +518,14 @@ class K8sTaskHandler extends TaskHandler implements FusionAwareTask {
         if( !task.isSuccess() )
             return
 
-        // k8s cluster will cleanup job on its own if ttl is set
+        // k8s cluster will cleanup job on its own if TTL is set
         if( useJobResource() && getPodOptions().getTtlSecondsAfterFinished() != null )
             return
 
+        delete0(podName)
+    }
+
+    private void delete0(String podName) {
         try {
             if ( useJobResource() )
                 client.jobDelete(podName)
@@ -532,15 +533,15 @@ class K8sTaskHandler extends TaskHandler implements FusionAwareTask {
                 client.podDelete(podName)
         }
         catch( Exception e ) {
-            log.warn "Unable to cleanup ${resourceType.lower()}: $podName -- see the log file for details", e
+            log.warn "Unable to delete ${resourceType.lower()}: $podName -- see the log file for details", e
         }
     }
 
-    private void determineNode(){
+    private void determineNode() {
         try {
             if ( k8sConfig.fetchNodeName() && !runsOnNode )
                 runsOnNode = client.getNodeOfPod( podName )
-        } catch ( Exception e ){
+        } catch ( Exception e ) {
             log.warn ("Unable to get the node name of pod $podName -- see the log file for details", e)
         }
     }

--- a/plugins/nf-k8s/src/test/nextflow/k8s/K8sTaskHandlerTest.groovy
+++ b/plugins/nf-k8s/src/test/nextflow/k8s/K8sTaskHandlerTest.groovy
@@ -497,8 +497,8 @@ class K8sTaskHandlerTest extends Specification {
         1 * handler.getState() >> fullState
         1 * handler.updateTimestamps(termState)
         1 * handler.readExitFile() >> EXIT_STATUS
-        1 * handler.deletePodIfSuccessful(task) >> null
-        1 * handler.savePodLogOnError(task) >> null
+        1 * handler.deleteJobIfSuccessful(task) >> null
+        1 * handler.saveJobLogOnError(task) >> null
         handler.task.exitStatus == EXIT_STATUS
         handler.task.@stdout == OUT_FILE
         handler.task.@stderr == ERR_FILE
@@ -528,8 +528,8 @@ class K8sTaskHandlerTest extends Specification {
         1 * handler.getState() >> [terminated: termState]
         1 * handler.updateTimestamps(termState)
         0 * handler.readExitFile()
-        1 * handler.deletePodIfSuccessful(task) >> null
-        1 * handler.savePodLogOnError(task) >> null
+        1 * handler.deleteJobIfSuccessful(task) >> null
+        1 * handler.saveJobLogOnError(task) >> null
         handler.task.exitStatus == 137
         handler.status == TaskStatus.COMPLETED
         result == true
@@ -770,19 +770,19 @@ class K8sTaskHandlerTest extends Specification {
         def TASK_FAIL = Mock(TaskRun); TASK_FAIL.isSuccess() >> false
 
         when:
-        handler.deletePodIfSuccessful(TASK_OK)
+        handler.deleteJobIfSuccessful(TASK_OK)
         then:
         1 * executor.getK8sConfig() >> new K8sConfig()
         1 * client.podDelete(POD_NAME) >> null
 
         when:
-        handler.deletePodIfSuccessful(TASK_OK)
+        handler.deleteJobIfSuccessful(TASK_OK)
         then:
         1 * executor.getK8sConfig() >> new K8sConfig(cleanup: true)
         1 * client.podDelete(POD_NAME) >> null
 
         when:
-        handler.deletePodIfSuccessful(TASK_FAIL)
+        handler.deleteJobIfSuccessful(TASK_FAIL)
         then:
         1 * executor.getK8sConfig() >> new K8sConfig(cleanup: false)
         0 * client.podDelete(POD_NAME) >> null
@@ -801,14 +801,14 @@ class K8sTaskHandlerTest extends Specification {
         def TASK_OK = Mock(TaskRun); TASK_OK.isSuccess() >> true
 
         when: 'job with ttlSecondsAfterFinished should not be deleted'
-        handler.deletePodIfSuccessful(TASK_OK)
+        handler.deleteJobIfSuccessful(TASK_OK)
         then:
         1 * executor.getK8sConfig() >> new K8sConfig()
         1 * handler.getPodOptions() >> new PodOptions([[ttlSecondsAfterFinished: 100]])
         0 * client.jobDelete(POD_NAME)
 
         when: 'job without ttlSecondsAfterFinished should be deleted'
-        handler.deletePodIfSuccessful(TASK_OK)
+        handler.deleteJobIfSuccessful(TASK_OK)
         then:
         1 * executor.getK8sConfig() >> new K8sConfig()
         1 * handler.getPodOptions() >> new PodOptions()
@@ -830,13 +830,13 @@ class K8sTaskHandlerTest extends Specification {
         def handler = Spy(new K8sTaskHandler(executor: executor, client: client, podName: POD_NAME))
 
         when:
-        handler.savePodLogOnError(task)
+        handler.saveJobLogOnError(task)
         then:
         task.isSuccess() >> true
         0 * client.podLog(_)
 
         when:
-        handler.savePodLogOnError(task)
+        handler.saveJobLogOnError(task)
         then:
         task.isSuccess() >> false
         task.getWorkDir() >> folder

